pkg/gc/gc_state_manager.go

```diff
@@ -543,6 +612,155 @@ func (m *GCStateManager) GetGCState(keyspaceID uint32) (GCState, error) {
 	return result, err
 }
 
+// GetAllKeyspacesGCStates returns the GC state of all keyspaces.
+// Returns a map from keyspaceID to GCState. Keyspaces without keyspace-level GC enabled will not be included.
+func (m *GCStateManager) GetAllKeyspacesGCStates() (map[uint32]GCState, error) {
+	// TODO: Handle the case that there are too many keyspaces and loading them at once is not suitable.
+	allKeyspaces, err := m.keyspaceManager.LoadRangeKeyspace(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// Do not guarantee atomicity among different keyspaces here.
+	results := make(map[uint32]GCState)
+	err = m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		nullKeyspaceState, err1 := m.getGCStateInTransaction(constant.NullKeyspaceID, wb)
+		if err1 != nil {
+			return err1
+		}
+		results[constant.NullKeyspaceID] = nullKeyspaceState
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, keyspaceMeta := range allKeyspaces {
+		if keyspaceMeta.Config[keyspace.GCManagementType] != keyspace.KeyspaceLevelGC {
+			results[keyspaceMeta.Id] = GCState{
+				KeyspaceID:      keyspaceMeta.Id,
+				IsKeyspaceLevel: false,
+			}
+			continue
+		}
+
+		err = m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+			state, err1 := m.getGCStateInTransaction(keyspaceMeta.Id, wb)
+			if err1 != nil {
+				return err1
+			}
+			results[keyspaceMeta.Id] = state
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return results, nil
+}
+
+// saturatingDuration returns a duration calculated by multiplying the given `ratio` and `base`, truncated within the
+// range [0, math.MaxInt64] to avoid negative value and overflowing.
+func saturatingDuration(ratio int64, base time.Duration) time.Duration {
```

Comment:
Why do we need this function?

Reply:
It's used for avoiding overflow after multiplication, so that it prevents the user from specifying a too large value and causing undefined behavior.